File: analysis/loneliness_exec_models.R
#!/usr/bin/env Rscript
#
# Loneliness × Executive Function analysis pipeline
# -------------------------------------------------
# This script ingests the exported CSV files produced by the Flutter tasks
# (participants info, survey scores, and cognitive summaries), constructs
# task-level and latent executive-function indicators, and runs the core
# frequentist tests requested in the IRB plan:
#   1) Does trait loneliness (UCLA v3) predict each EF readout after
#      controlling for DASS-21 Depression/Anxiety/Stress?
#   2) Does loneliness load onto a shared “meta-control” factor spanning
#      Stroop, PRP, and WCST metrics?
# The script prints model summaries to the console and saves tidy p-value
# tables to `results/analysis_outputs/`.
#
# Usage (from repo root):
#   Rscript analysis/loneliness_exec_models.R
# -------------------------------------------------

suppressPackageStartupMessages({
  pkgs <- c("tidyverse", "janitor", "broom", "psych", "glue")
  need <- pkgs[!pkgs %in% installed.packages()[, "Package"]]
  if (length(need) > 0) install.packages(need)
  lapply(pkgs, library, character.only = TRUE)
})

data_dir <- "results"
output_dir <- file.path(data_dir, "analysis_outputs")
if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)

message("1) Loading CSV exports …")
participants <- read_csv(file.path(data_dir, "1_participants_info.csv"),
                         show_col_types = FALSE) |>
  clean_names()

surveys <- read_csv(file.path(data_dir, "2_surveys_results.csv"),
                    show_col_types = FALSE) |>
  clean_names()

cog <- read_csv(file.path(data_dir, "3_cognitive_tests_summary.csv"),
                show_col_types = FALSE) |>
  clean_names()

message(glue("   • Participants: {nrow(participants)} rows"))
message(glue("   • Survey rows : {nrow(surveys)} rows"))
message(glue("   • Cognitive   : {nrow(cog)} rows"))

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
zscore <- function(x) {
  if (all(is.na(x))) return(rep(NA_real_, length(x)))
  if (sd(x, na.rm = TRUE) == 0) return(rep(0, length(x)))
  as.numeric(scale(x))
}

clean_gender <- function(x) {
  case_when(
    is.na(x) ~ NA_character_,
    str_detect(x, "\uB0A8") ~ "male",   # Korean char U+B0A8 ("nam")
    str_detect(x, "\uC5EC") ~ "female", # Korean char U+C5EC ("yeo")
    TRUE ~ NA_character_
  )
}

# ---------------------------------------------------------------------------
# Surveys: derive UCLA trait loneliness + DASS state mood controls
# ---------------------------------------------------------------------------
ucla <- surveys |>
  filter(str_to_lower(survey_name) == "ucla") |>
  transmute(participant_id, ucla_total = score)

dass <- surveys |>
  filter(str_to_lower(survey_name) == "dass") |>
  transmute(
    participant_id,
    dass_anx = score_a,
    dass_stress = score_s,
    dass_dep = score_d
  ) |>
  mutate(dass_total = dass_anx + dass_stress + dass_dep)

# ---------------------------------------------------------------------------
# Cognitive summaries → task-specific EF indicators
# ---------------------------------------------------------------------------
prp <- cog |>
  filter(str_to_lower(test_name) == "prp") |>
  transmute(
    participant_id,
    prp_n_trials = n_trials,
    prp_acc_t1 = acc_t1,
    prp_acc_t2 = acc_t2,
    prp_mrt_t1 = mrt_t1,
    prp_mrt_t2 = mrt_t2,
    prp_rt_short = rt2_soa_50,
    prp_rt_long = rt2_soa_1200,
    prp_bottleneck = rt2_soa_50 - rt2_soa_1200,
    prp_rt_slope = (rt2_soa_50 - rt2_soa_1200) / 1150
  )

stroop <- cog |>
  filter(str_to_lower(test_name) == "stroop") |>
  transmute(
    participant_id,
    stroop_total_trials = total,
    stroop_accuracy = accuracy,
    stroop_effect = stroop_effect,
    stroop_mrt_incong = mrt_incong,
    stroop_mrt_cong = mrt_cong,
    stroop_mrt_total = mrt_total
  )

wcst <- cog |>
  filter(str_to_lower(test_name) == "wcst") |>
  transmute(
    participant_id,
    wcst_total_errors = total_error_count,
    wcst_persev_errors = perseverative_error_count,
    wcst_nonpersev_errors = non_perseverative_error_count,
    wcst_completed_categories = completed_categories,
    wcst_conceptual_pct = conceptual_level_responses_percent,
    wcst_persev_resp_pct = perseverative_responses_percent,
    wcst_failure_to_maintain_set = failure_to_maintain_set
  )

# ---------------------------------------------------------------------------
# Merge everything
# ---------------------------------------------------------------------------
analysis_df <- participants |>
  select(participant_id, age, gender) |>
  mutate(
    age = as.numeric(age),
    gender = factor(clean_gender(gender))
  ) |>
  left_join(ucla, by = "participant_id") |>
  left_join(dass, by = "participant_id") |>
  left_join(prp, by = "participant_id") |>
  left_join(stroop, by = "participant_id") |>
  left_join(wcst, by = "participant_id")

message(glue("2) Combined analytic dataset: {nrow(analysis_df)} participants"))

# Z-scores for focal predictors
analysis_df <- analysis_df |>
  mutate(
    z_ucla = zscore(ucla_total),
    z_dass_dep = zscore(dass_dep),
    z_dass_anx = zscore(dass_anx),
    z_dass_stress = zscore(dass_stress)
  )

# ---------------------------------------------------------------------------
# Latent meta-control factor (1-factor PCA on EF metrics)
# ---------------------------------------------------------------------------
ef_vars <- c("stroop_effect", "prp_bottleneck", "wcst_total_errors")
ef_complete <- analysis_df |>
  select(participant_id, all_of(ef_vars)) |>
  drop_na()

if (nrow(ef_complete) >= 15) {
  message("3) Estimating 1-factor PCA (meta-control) …")
  pca_fit <- psych::principal(scale(ef_complete |> select(-participant_id)),
                              nfactors = 1, rotate = "none", scores = TRUE)
  meta_scores <- tibble(
    participant_id = ef_complete$participant_id,
    meta_control = as.numeric(pca_fit$scores[, 1])
  )
  analysis_df <- analysis_df |>
    left_join(meta_scores, by = "participant_id")
  write_csv(
    tibble(
      indicator = ef_vars,
      loading = as.numeric(pca_fit$loadings[, 1])
    ),
    file.path(output_dir, "meta_control_loadings.csv")
  )
  message("   • PCA loadings saved to meta_control_loadings.csv")
} else {
  message("3) PCA skipped (need ≥15 complete cases).")
  analysis_df$meta_control <- NA_real_
}

# ---------------------------------------------------------------------------
# Frequentist models: Loneliness predicting EF outcomes
# ---------------------------------------------------------------------------
model_specs <- tribble(
  ~outcome, ~nice_name,
  "stroop_effect", "Stroop interference (ms)",
  "prp_bottleneck", "PRP bottleneck (short-long RT)",
  "wcst_total_errors", "WCST total errors",
  "meta_control", "Latent meta-control factor"
)

fit_lm <- function(dep, data) {
  data_mod <- data |>
    select(
      !!sym(dep), z_ucla, z_dass_dep, z_dass_anx, z_dass_stress,
      age, gender
    ) |>
    rename(y = !!sym(dep)) |>
    drop_na(y, z_ucla, z_dass_dep, z_dass_anx, z_dass_stress)
  if (nrow(data_mod) < 20) return(NULL)
  lm(y ~ z_ucla + z_dass_dep + z_dass_anx + z_dass_stress +
       age + gender,
     data = data_mod)
}

message("4) Running hierarchical linear models …")
model_tbl <- model_specs |>
  mutate(
    model = map(outcome, ~fit_lm(.x, analysis_df)),
    valid = map_lgl(model, ~!is.null(.x))
  ) |>
  filter(valid)

if (nrow(model_tbl) == 0) {
  stop("No models could be estimated (insufficient complete cases).")
}

coef_tbl <- model_tbl |>
  mutate(
    coefs = map2(model, nice_name, ~tidy(.x, conf.int = TRUE) |>
                   mutate(outcome = .y))
  ) |>
  select(coefs) |>
  unnest(coefs) |>
  relocate(outcome, term)

fit_tbl <- model_tbl |>
  mutate(
    glance = map2(model, nice_name, ~glance(.x) |> mutate(outcome = .y))
  ) |>
  select(glance) |>
  unnest(glance) |>
  relocate(outcome, r.squared, adj.r.squared, p.value, AIC, BIC, nobs)

write_csv(coef_tbl, file.path(output_dir, "loneliness_models_coefficients.csv"))
write_csv(fit_tbl, file.path(output_dir, "loneliness_models_fit.csv"))

message("   • Coefficients saved to loneliness_models_coefficients.csv")
message("   • Model fit stats saved to loneliness_models_fit.csv")

cat("\n=== Key p-values for UCLA Loneliness ===\n")
coef_tbl |>
  filter(term == "z_ucla") |>
  arrange(p.value) |>
  mutate(across(c(estimate, std.error, conf.low, conf.high, p.value),
                ~round(., 4))) |>
  print(n = Inf)

message("\nDone.")

File: analysis/loneliness_exec_models.py
"""
Python implementation of the Loneliness × Executive Function analysis.

Replicates the R workflow in a pure Python stack so it can be executed
directly here. Steps:
1. Load participants, survey, and cognitive-summary CSV exports.
2. Derive task-level EF metrics (Stroop, PRP, WCST) and survey predictors.
3. Merge into a single analysis DataFrame with z-scored loneliness/DASS.
4. Optional: run 1-factor PCA across EF indicators to extract a latent
   "meta-control" score, saving loadings if ≥15 complete cases.
5. Fit OLS models for each EF outcome with loneliness + DASS controls,
   plus age and gender covariates (requiring ≥20 complete rows per model).
6. Save coefficient and model-fit tables to results/analysis_outputs/.

Usage:
    python analysis/loneliness_exec_models.py
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional

import numpy as np
import pandas as pd
from data_loader_utils import load_master_dataset
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import statsmodels.formula.api as smf

from data_loader_utils import normalize_gender_series


# ---------------------------------------------------------------------------
# Data loading helpers
# ---------------------------------------------------------------------------
BASE_DIR = Path(__file__).resolve().parent.parent
RESULTS_DIR = BASE_DIR / "results"
OUTPUT_DIR = RESULTS_DIR / "analysis_outputs"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


COLUMN_RENAMES = {
    "participantid": "participant_id",
    "surveyname": "survey_name",
    "testname": "test_name",
    "conceptuallevelresponses": "conceptual_level_responses",
    "nonperseverativeerrorcount": "non_perseverative_error_count",
    "perseverativeresponses": "perseverative_responses",
    "trialstofirstconceptualresp": "trials_to_first_conceptual_resp",
    "totaltrialcount": "total_trial_count",
    "totalcorrectcount": "total_correct_count",
    "conceptuallevelresponsespercent": "conceptual_level_responses_percent",
    "completedcategories": "completed_categories",
    "totalerrorcount": "total_error_count",
    "perseverativeerrorcount": "perseverative_error_count",
    "trialstocompletefirstcategory": "trials_to_complete_first_category",
    "learningtolearn": "learning_to_learn",
    "perseverativeresponsespercent": "perseverative_responses_percent",
    "failuretomaintainset": "failure_to_maintain_set",
    "learningtolearnheatonclrdelta": "learning_to_learn_heaton_clr_delta",
    "learningefficiencydeltatrials": "learning_efficiency_delta_trials",
    "hasfirstclr": "has_first_clr",
    "trialstofirstconceptualresp0": "trials_to_first_conceptual_resp0",
    "categoryclrpercents": "category_clr_percents",
}


def read_csv_lower(path: Path) -> pd.DataFrame:
    """Read CSV with pandas and standardize column names to snake_case-ish."""
    df = pd.read_csv(path)
    df.columns = (
        df.columns.str.strip()
        .str.lower()
        .str.replace(" ", "_")
        .str.replace(".", "_", regex=False)
    )
    df = df.rename(columns=COLUMN_RENAMES)
    return df


# ---------------------------------------------------------------------------
# Feature engineering
# ---------------------------------------------------------------------------
def zscore(series: pd.Series) -> pd.Series:
    if series.std(skipna=True) == 0 or series.dropna().empty:
        return pd.Series([np.nan] * len(series), index=series.index)
    return (series - series.mean(skipna=True)) / series.std(skipna=True)


def build_analysis_dataframe() -> pd.DataFrame:
    master = load_master_dataset(use_cache=True, force_rebuild=True, merge_cognitive_summary=True)
    print(f"[AUDIT] Master dataset loaded with {len(master)} rows.")

    def _pick_column(df: pd.DataFrame, candidates: List[str]) -> pd.Series:
        for c in candidates:
            if c in df.columns:
                return df[c]
        return pd.Series(np.nan, index=df.index)

    analysis = master.loc[:, ["participant_id", "age", "gender_normalized"]].rename(columns={"gender_normalized": "gender"})
    analysis["gender"] = normalize_gender_series(analysis["gender"])
    analysis["age"] = pd.to_numeric(analysis["age"], errors="coerce")

    # Core predictors
    if "ucla_total" in master.columns:
        analysis["ucla_total"] = master["ucla_total"]
    elif "ucla_score" in master.columns:
        analysis["ucla_total"] = master["ucla_score"]

    analysis["dass_dep"] = master.get("dass_depression")
    analysis["dass_anx"] = master.get("dass_anxiety")
    analysis["dass_stress"] = master.get("dass_stress")
    if {"dass_dep", "dass_anx", "dass_stress"} <= set(analysis.columns):
        analysis["dass_total"] = analysis[["dass_dep", "dass_anx", "dass_stress"]].sum(axis=1)

    # EF metrics sourced from standardized summaries (PRP/Stroop/WCST)
    analysis["prp_bottleneck"] = master.get("prp_bottleneck")
    if "stroop_interference" in master.columns:
        analysis["stroop_effect"] = master["stroop_interference"]
    elif {"rt_mean_incongruent", "rt_mean_congruent"} <= set(master.columns):
        analysis["stroop_effect"] = master["rt_mean_incongruent"] - master["rt_mean_congruent"]

    # WCST errors from cognitive summary (merged raw columns kept as-is)
    analysis["wcst_total_errors"] = _pick_column(master, ["wcst_total_errors", "total_error_count", "totalErrorCount"])
    analysis["wcst_persev_errors"] = _pick_column(master, ["wcst_persev_errors", "perseverative_error_count", "perseverativeErrorCount"])
    analysis["wcst_nonpersev_errors"] = _pick_column(master, ["wcst_nonpersev_errors", "non_perseverative_error_count", "nonPerseverativeErrorCount"])

    # Optional WCST auxiliaries (kept if present)
    analysis["wcst_persev_resp_pct"] = _pick_column(master, ["wcst_persev_resp_pct", "perseverative_responses_percent", "perseverativeResponsesPercent"])
    analysis["wcst_completed_categories"] = _pick_column(master, ["wcst_completed_categories", "completed_categories", "completedCategories"])
    analysis["wcst_conceptual_pct"] = _pick_column(master, ["wcst_conceptual_pct", "conceptual_level_responses_percent", "conceptualLevelResponsesPercent"])
    analysis["wcst_failure_to_maintain_set"] = _pick_column(master, ["wcst_failure_to_maintain_set", "failure_to_maintain_set", "failureToMaintainSet"])

    # Add z-scores
    for col, new_col in [
        ("ucla_total", "z_ucla"),
        ("dass_dep", "z_dass_dep"),
        ("dass_anx", "z_dass_anx"),
        ("dass_stress", "z_dass_stress"),
    ]:
        analysis[new_col] = zscore(analysis[col])

    return analysis


# ---------------------------------------------------------------------------
# PCA meta-control factor
# ---------------------------------------------------------------------------
def add_meta_control(df: pd.DataFrame) -> pd.DataFrame:
    ef_cols = ["stroop_effect", "prp_bottleneck", "wcst_total_errors"]
    complete = df.dropna(subset=ef_cols).copy()
    if len(complete) < 15:
        print("PCA skipped (need at least 15 complete EF rows).")
        df["meta_control"] = np.nan
        return df

    scaler = StandardScaler()
    scaled = scaler.fit_transform(complete[ef_cols])
    pca = PCA(n_components=1)
    scores = pca.fit_transform(scaled).ravel()
    complete = complete.assign(meta_control=scores)
    df = df.merge(complete[["participant_id", "meta_control"]], on="participant_id", how="left")

    loadings = pd.DataFrame(
        {
            "indicator": ef_cols,
            "loading": pca.components_[0],
        }
    )
    loadings.to_csv(OUTPUT_DIR / "meta_control_loadings.csv", index=False)
    print("Saved PCA loadings to meta_control_loadings.csv")
    return df


# ---------------------------------------------------------------------------
# Regression modeling
# ---------------------------------------------------------------------------
@dataclass
class ModelResult:
    outcome: str
    nobs: int
    r_squared: float
    adj_r_squared: float
    aic: float
    bic: float
    p_value: float
    coefficients: pd.DataFrame


def run_models(df: pd.DataFrame) -> List[ModelResult]:
    specs = [
        ("stroop_effect", "Stroop interference (ms)"),
        ("prp_bottleneck", "PRP bottleneck (short-long RT)"),
        ("wcst_total_errors", "WCST total errors"),
        ("meta_control", "Latent meta-control factor"),
    ]
    results: List[ModelResult] = []
    formula_template = "y ~ z_ucla + z_dass_dep + z_dass_anx + z_dass_stress + age + C(gender)"

    for outcome, nice in specs:
        cols_needed = [
            outcome,
            "z_ucla",
            "z_dass_dep",
            "z_dass_anx",
            "z_dass_stress",
            "age",
            "gender",
        ]
        data = df[cols_needed].dropna()
        if len(data) < 20:
            print(f"Skipping {outcome}: only {len(data)} complete rows.")
            continue

        data = data.rename(columns={outcome: "y"})
        model = smf.ols(formula=formula_template, data=data).fit(cov_type="HC3")
        coef_raw = model.summary2().tables[1].copy()
        # Harmonize column names across robust/non-robust variants
        rename_map = {"Coef.": "estimate", "Std.Err.": "std_error", "[0.025": "conf_low", "0.975]": "conf_high"}
        coef_raw = coef_raw.rename(columns=rename_map)
        if "P>|t|" in coef_raw.columns:
            coef_raw = coef_raw.rename(columns={"P>|t|": "p_value", "t": "stat"})
        if "P>|z|" in coef_raw.columns:
            coef_raw = coef_raw.rename(columns={"P>|z|": "p_value", "z": "stat"})
        coef_df = coef_raw.reset_index().rename(columns={"index": "term"})

        # Use a robust Wald test (excluding intercept) for omnibus p-value
        omnibus_p = np.nan
        try:
            if len(model.params) > 1:
                R = np.eye(len(model.params))[1:]
                wald = model.wald_test(R, use_f=True, scalar=False)
                omnibus_p = float(wald.pvalue)
        except Exception:
            pass

        results.append(
            ModelResult(
                outcome=nice,
                nobs=int(model.nobs),
                r_squared=float(model.rsquared),
                adj_r_squared=float(model.rsquared_adj),
                aic=float(model.aic),
                bic=float(model.bic),
                p_value=omnibus_p,
                coefficients=coef_df,
            )
        )
        print(f"Fitted model for {nice} (n={int(model.nobs)})")

    return results


def save_outputs(results: List[ModelResult]) -> None:
    if not results:
        print("No models were fit; nothing to save.")
        return

    coef_rows = []
    fit_rows = []
    for res in results:
        coef = res.coefficients.copy()
        coef.insert(0, "outcome", res.outcome)
        coef_rows.append(coef)

        fit_rows.append(
            {
                "outcome": res.outcome,
                "r_squared": res.r_squared,
                "adj_r_squared": res.adj_r_squared,
                "AIC": res.aic,
                "BIC": res.bic,
                "p_value": res.p_value,
                "nobs": res.nobs,
            }
        )

    coef_df = pd.concat(coef_rows, ignore_index=True)
    # Add FDR (BH) q-values across all coefficients
    try:
        from statsmodels.stats.multitest import multipletests
        if "p_value" in coef_df.columns:
            coef_df["q_value"] = multipletests(coef_df["p_value"].astype(float).values, method="fdr_bh")[1]
    except Exception:
        pass
    fit_df = pd.DataFrame(fit_rows)

    coef_path = OUTPUT_DIR / "loneliness_models_coefficients_py.csv"
    fit_path = OUTPUT_DIR / "loneliness_models_fit_py.csv"
    coef_df.to_csv(coef_path, index=False)
    fit_df.to_csv(fit_path, index=False)
    print(f"Saved coefficients to {coef_path.name}")
    print(f"Saved fit stats to {fit_path.name}")

    lon_rows = coef_df.query("term == 'z_ucla'")
    if not lon_rows.empty:
        print("\n=== Key p-values for UCLA Loneliness (Python) ===")
        print(lon_rows[["outcome", "estimate", "std_error", "conf_low", "conf_high", "p_value"]])


def main():
    analysis_df = build_analysis_dataframe()
    analysis_df = add_meta_control(analysis_df)
    results = run_models(analysis_df)
    save_outputs(results)


if __name__ == "__main__":
    main()
